constrain-lib/src/test/groovy/org/gradle/dependency/constrain/lib/serialize/ConstraintsJsonReaderTest.groovy

```diff
@@ -0,0 +1,420 @@
+  - Remove constraint at position 0:
+    - {"group":"com.a", "name":"aaa", "suggestedVersion":"1.0.1"}
+  - Insert constraint at position 2:
+    - {"group":"com.a", "name":"aaa", "suggestedVersion":"1.0.1"}
+""".trim()
+        }
+    }
+
+    def "fails with error when sequences of constraints aren't sorted lexicographically by group:name"() {
+        when:
+        parse """
+{
+  "version": "1.0.0",
+  "dependencyConstraints": [
+    {
+      "groupId": "com.c",
+      "name": "aaa",
+      "suggestedVersion": "1.0.0",
+      "because": {
+        "reason": "Reason"
+      }
+    },
+    {
+      "groupId": "com.d",
+      "name": "aaa",
+      "suggestedVersion": "1.0.0",
+      "because": {
+        "reason": "Reason"
+      }
+    },
+    {
+      "groupId": "com.a",
+      "name": "aaa",
+      "suggestedVersion": "1.0.0",
+      "because": {
+        "reason": "Reason"
+      }
+    },
+    {
+      "groupId": "com.b",
+      "name": "aaa",
+      "suggestedVersion": "1.0.0",
+      "because": {
+        "reason": "Reason"
+      }
+    }
+  ]
+}
+"""
+        then:
+        def ex = thrown(DependencyConstrainException)
+        verifyAll(ex) {
+            getMessage() == "Unable to read dependency constraints"
+            cause instanceof DependencyConstrainException
+            cause.getMessage() == """
+Constrains were not sorted by group:name:suggestedVersion in lexicographical order:
+  - Remove constraints at position 0 through 1:
+    - {"group":"com.c", "name":"aaa", "suggestedVersion":"1.0.0"}
+    - {"group":"com.d", "name":"aaa", "suggestedVersion":"1.0.0"}
+  - Insert constraints at position 4 through 5:
+    - {"group":"com.c", "name":"aaa", "suggestedVersion":"1.0.0"}
+    - {"group":"com.d", "name":"aaa", "suggestedVersion":"1.0.0"}
+""".trim()
+        }
+    }
+
+    def "parsing input not formatting throws an error"() {
+        when:
+        parse """
+{
+  "version": "1.0.0",
+  "dependencyConstraints": [{
+      "groupId": "com.c",
+          "name": "aaa",
+      "suggestedVersion": "1.0.0",
+      "because": { "reason": "Reason"
+      }
+    }
+    ]
+    }
+"""
+        then:
+        def ex = thrown(DependencyConstrainException)
+        verifyAll(ex) {
+            getMessage() == "Unable to read dependency constraints"
+            cause instanceof DependencyConstrainException
+            cause.getMessage() == """
+The dependency constraints are not formatted correctly. Please apply this patch to fix the formatting:
+  --- gradle/dependency-constraints.json
+  +++ gradle/dependency-constraints.json
+  @@ -3,1 +3,2 @@
+  -  "dependencyConstraints": [{
+  +  "dependencyConstraints": [
+  +    {
+  @@ -5,1 +6,1 @@
+  -          "name": "aaa",
+  +      "name": "aaa",
+  @@ -7,1 +8,2 @@
+  -      "because": { "reason": "Reason"
+  +      "because": {
+  +        "reason": "Reason"
+  @@ -10,2 +12,2 @@
+  -    ]
+  -    }
+  +  ]
+  +}
+""".trim()
```

Comment:
I'm a pretty big fan of being able to not just tell the user that the file they specified is incorrectly formatted, but also provide the diff to make to the file to actually make it match the expected formatting.

Additionally, you can copy this text to your clipboard, and IntelliJ has a feature `Git > Patch > Apply Patch From Clipboard` which makes it super easy to fix this file when you get this error.